framework/include/actions/ActionWarehouse.h

```diff
@@ -101,6 +101,64 @@ class ActionWarehouse :
   const std::vector<Action *> & getActionsByName(const std::string & task) const;
 
   /**
+   * Retrieve an action with its name and the desired type.
+   * @param name The action name.
+   */
+  template <class T>
+  const T & getAction(const std::string & name)
+    {
+      T* p = NULL;
+      for (std::vector<std::string>::iterator it = _ordered_names.begin(); it != _ordered_names.end(); ++it)
+      {
+        std::string task = *it;
+        for (ActionIterator act_iter = actionBlocksWithActionBegin(task);
+             act_iter != actionBlocksWithActionEnd(task);
+             ++act_iter)
+        {
+          if ((*act_iter)->name() == name)
+          {
+            p = dynamic_cast<T*>(*act_iter);
+            if (!p)
+              mooseError("The action with name being "+name+" is not in the desired type");
+            else
+              break;
+          }
+        }
+      }
+      if (!p)
+        mooseError("Action with name being "+name+" does not exist");
+      return *p;
+    }
+
+  /**
+   * Retrieve all actions in a specific type.
+   */
+  template <class T>
+  std::vector<const T *> getActions()
```

Comment:
`getActionByName()` should be renamed if you plan to actually implement the proper by name method. Leaving it incorrect will create confusion. Also, if you are in the wearhouse, we already have a vector of `Actions`. You should be able to iterate through the vector instead of iterating through the tasks to get the Actions putting them in a set only to get the Actions you could have retrieved from the vector.